types/jest/index.d.ts

```diff
@@ -110,19 +113,15 @@ declare namespace jest {
     /**
      * Creates a mock function. Optionally takes a mock implementation.
      */
-    function fn<T extends {}>(implementation: (...args: any[]) => T): Mock<T>;
-    /**
-     * Creates a mock function. Optionally takes a mock implementation.
-     */
-    function fn<T>(implementation?: (...args: any[]) => any): Mock<T>;
+    function fn<T, Y extends any[]>(implementation?: (...args: Y) => T): Mock<T, Y>;
```

Comment:
You need `Y extends never[]`, otherwise this refuses functions that have `never` parameters (rare, but possible).

```ts
declare function unreachable(arg: never): never

jest.fn(unreachable) // will error with `any[]`
```

Reply:
Actually, is there any reason why this is not just `function fn<T extends (...args: never[]) => unknown>(implementation?: T): Mock<T>`, with `Mock` changed to hold a function generic instead?

It makes declaring the `Mock` methods more annoying _but_ it makes using `.fn` a lot easier when you want to make sure your mock function conforms to the interface of a function that already exists. Just call `jest.fn<typeof myfunc>(/* inline function will even get contextual params/return types */)`.

Reply:
Also, the default type (when implementation is not specified) should be `(...args: any[]) => void`. Perhaps it should be on a 0 argument overload instead of an optional argument to avoid misuse.

Reply:
> You need `Y extends never[]`, otherwise this refuses functions that have `never` parameters (rare, but possible).
> 
> ```ts
> declare function unreachable(arg: never): never
> 
> jest.fn(unreachable) // will error with `any[]`
> ```
This is false. I just updated the tests and it doesn't error. 



> Actually, is there any reason why this is not just `function fn<T extends (...args: never[]) => unknown>(implementation?: T): Mock<T>`, with `Mock` changed to hold a function generic instead?
> 
> It makes declaring the `Mock` methods more annoying _but_ it makes using `.fn` a lot easier when you want to make sure your mock function conforms to the interface of a function that already exists. Just call `jest.fn<typeof myfunc>(/* inline function will even get contextual params/return types */)`.

Because this would be a breaking change. By adding a new generic with a default value, existing code will not be affected.



> Also, the default type (when implementation is not specified) should be `(...args: any[]) => void`. Perhaps it should be on a 0 argument overload instead of an optional argument to avoid misuse.

It should not return void. This should compile:
```
jest.fn().mockImplementation((test: number) => test);
```

Reply:
Hmm... I need to brush up on the `...args: never[]` thing, then. I had seen it as a recommendation from the TS team *somewhere* a few months ago.

As for the second one, that's because you're specifying a mock implementation after the `.fn` call but... that is supported API because JS is weird like that 😵

If you use `.fn()` _without_ also adding a more specific `.mock*` it'll return `void` by default. But currently methods cannot alter the type of anything on their `this`...

Reply:
FYI opened a new PR #32579

Reply:
@antoinebrault Can you help me understand how to use the new:

```ts
function fn<T, Y extends any[]>(implementation: (...args: Y) => T): Mock<T, Y>;
```

I understand the T, but what is expected for the Y. After updating to this version, several of the mockImplementations, are throwing typing issues.

Reply:
@josecolella Y is the arguments types of the implementation function.

e.g.
```typescript
function x(y: number) {
  return y + '';
}
jest.fn(x) // will be Mock<string, [number]>
```

When mocking implementations, you have to use the same arguments types as the original function, but arguments are optional.
e.g. 
```typescript
function x(y: number) {
  return y + '';
}
jest.fn(x).mockImplementation(() => ''); // will work
```

Reply:
@antoinebrault What about for mockImplementation of classes? As of now it's complaining that I need to implement all the methods within the class, when in reality only one method needs to be mocked. 

```ts
interface SomeInterface {
      foo() 
      foo2()
}

jest.fn<SomeInterface>(() => ({
 foo: jest.fn().mockReturnValue(true)
}));
```

With new types this no longer works. Do you have an example of how to use this new version with class mock implementations?

Reply:
> @antoinebrault What about for mockImplementation of classes? As of now it's complaining that I need to implement all the methods within the class, when in reality only one method needs to be mocked.
> 
you can use Partial for that usecase.  ```jest.Mocked<Partial<Type>>```

> ```ts
> interface SomeInterface {
>       foo() 
>       foo2()
> }
> 
> jest.fn<SomeInterface>(() => ({
>  foo: jest.fn().mockReturnValue(true)
> }));
> ```
> 
> With new types this no longer works. Do you have an example of how to use this new version with class mock implementations?

```typescript
interface SomeInterface {
    foo(): string,
    foo2(): number
}
const y: jest.Mocked<SomeInterface> = {
    foo: jest.fn().mockReturnValue(''),
    foo2: jest.fn().mockReturnValue(1)
};
```

Reply:
@antoinebrault Thanks for your explanation and examples.